File: rust/src/generate.rs
// SPDX-License-Identifier: CC-BY-4.0
// Copyright (c) 2025-2026 fumi-engineer

//! Token generation with sampling strategies.
//!
//! Autoregressive generation: at each step, run full forward pass on all
//! tokens so far, take the logits for the last position, sample a token,
//! append it, and repeat. (No KV cache -- educational implementation.)
//!
//! Sampling strategies:
//! - **Greedy**: argmax(logits)
//! - **Temperature**: softmax(logits / T), then sample from distribution
//! - **Top-k**: zero out all but the top-k logits, then temperature sample
//! - **Top-p (nucleus)**: keep smallest set of tokens with cumulative prob >= p

use std::cmp::Ordering;

use crate::model::MoETransformer;
use crate::tensor;

/// Sampling strategy for token generation.
pub enum SamplingStrategy {
    /// Always pick the highest-probability token
    Greedy,
    /// Sample from softmax(logits / temperature)
    Sample { temperature: f32 },
    /// Keep only top-k logits, then temperature sample
    TopK { k: usize, temperature: f32 },
    /// Keep tokens until cumulative probability >= top_p, then temperature sample
    TopP { top_p: f32, temperature: f32 },
}

pub fn pick_token(logits: &[f32], strategy: &SamplingStrategy, state: &mut u64) -> usize {
    match *strategy {
        SamplingStrategy::Greedy => argmax(logits),
        SamplingStrategy::Sample { temperature } => sample_from_logits(logits, temperature, state),
        SamplingStrategy::TopK { k, temperature } => sample_top_k(logits, k, temperature, state),
        SamplingStrategy::TopP { top_p, temperature } => {
            sample_top_p(logits, top_p, temperature, state)
        }
    }
}

/// Autoregressive generation loop.
///
/// No KV cache: re-runs the full sequence through the model at each step.
/// This is O(n^2 * model_cost) but simple and correct for educational purposes.
///
/// Logits layout: [1, seq_len, vocab_size] flattened. The last position's logits
/// are at offset (seq_len-1)*vocab .. seq_len*vocab.
pub fn generate(
    model: &mut MoETransformer,
    prompt: &[usize],
    max_len: usize,
    strategy: &SamplingStrategy,
    seed: u64,
) -> Vec<usize> {
    let mut tokens = if prompt.is_empty() {
        vec![0]
    } else {
        prompt.to_vec()
    };
    let mut state = seed;
    let vocab = model.config().vocab_size;

    while tokens.len() < max_len {
        // Full forward pass on entire sequence (no KV cache)
        let logits = model.forward_ids(&tokens, 1, tokens.len());
        // Extract logits for the last position only
        let last = &logits.data()[(tokens.len() - 1) * vocab..tokens.len() * vocab];
        tokens.push(pick_token(last, strategy, &mut state));
    }

    tokens
}

pub fn generate_greedy(model: &mut MoETransformer, prompt: &[usize], max_len: usize) -> Vec<usize> {
    generate(model, prompt, max_len, &SamplingStrategy::Greedy, 0)
}

pub fn generate_sample(
    model: &mut MoETransformer,
    prompt: &[usize],
    max_len: usize,
    temperature: f32,
    seed: u64,
) -> Vec<usize> {
    generate(
        model,
        prompt,
        max_len,
        &SamplingStrategy::Sample { temperature },
        seed,
    )
}

pub fn generate_top_k(
    model: &mut MoETransformer,
    prompt: &[usize],
    max_len: usize,
    k: usize,
    temperature: f32,
    seed: u64,
) -> Vec<usize> {
    generate(
        model,
        prompt,
        max_len,
        &SamplingStrategy::TopK { k, temperature },
        seed,
    )
}

pub fn generate_top_p(
    model: &mut MoETransformer,
    prompt: &[usize],
    max_len: usize,
    top_p: f32,
    temperature: f32,
    seed: u64,
) -> Vec<usize> {
    generate(
        model,
        prompt,
        max_len,
        &SamplingStrategy::TopP { top_p, temperature },
        seed,
    )
}

fn argmax(slice: &[f32]) -> usize {
    slice
        .iter()
        .enumerate()
        .max_by(|a, b| a.1.partial_cmp(b.1).unwrap_or(Ordering::Equal))
        .map(|(i, _)| i)
        .unwrap_or(0)
}

fn sorted_indices_desc(values: &[f32]) -> Vec<usize> {
    let mut idxs: Vec<usize> = (0..values.len()).collect();
    idxs.sort_unstable_by(|&a, &b| values[b].partial_cmp(&values[a]).unwrap_or(Ordering::Equal));
    idxs
}

/// Temperature sampling: softmax(logits / T), then categorical sample.
///
/// p_i = exp(logit_i / T) / sum_j(exp(logit_j / T))
/// T -> 0: approaches greedy. T -> inf: approaches uniform.
fn sample_from_logits(logits: &[f32], temperature: f32, state: &mut u64) -> usize {
    if temperature <= 0.0 {
        return argmax(logits);
    }
    // Multiply by 1/T instead of dividing each element (one division vs N)
    let inv_temp = 1.0 / temperature;
    let mut probs: Vec<f32> = logits.iter().map(|x| x * inv_temp).collect();
    tensor::softmax_in_place(&mut probs);
    sample_from_probs(&probs, state)
}

/// Top-k sampling: keep only the top-k logits, set rest to -inf, then sample.
/// Setting non-top-k logits to -inf means softmax gives them probability 0.
fn sample_top_k(logits: &[f32], k: usize, temperature: f32, state: &mut u64) -> usize {
    if k == 0 || k >= logits.len() {
        return sample_from_logits(logits, temperature, state);
    }
    let idxs = sorted_indices_desc(logits);
    let mut filtered = vec![f32::NEG_INFINITY; logits.len()];
    for &idx in &idxs[..k] {
        filtered[idx] = logits[idx];
    }
    sample_from_logits(&filtered, temperature, state)
}

/// Top-p (nucleus) sampling: keep the smallest set of tokens whose cumulative
/// probability exceeds top_p, then renormalize and sample.
///
/// Adaptive: high-confidence predictions keep fewer tokens (sharper distribution),
/// uncertain predictions keep more (flatter distribution).
fn sample_top_p(logits: &[f32], top_p: f32, temperature: f32, state: &mut u64) -> usize {
    if !(0.0..1.0).contains(&top_p) {
        return sample_from_logits(logits, temperature, state);
    }
    if temperature <= 0.0 {
        return argmax(logits);
    }
    let inv_temp = 1.0 / temperature;
    let mut probs: Vec<f32> = logits.iter().map(|x| x * inv_temp).collect();
    tensor::softmax_in_place(&mut probs);

    // Sort by probability descending, accumulate until >= top_p
    let idxs = sorted_indices_desc(&probs);
    let mut kept = Vec::new();
    let mut mass = 0.0;
    for idx in idxs {
        kept.push(idx);
        mass += probs[idx];
        if mass >= top_p {
            break;
        }
    }

    // Renormalize the kept tokens and sample
    let mut truncated: Vec<f32> = kept.iter().map(|&idx| probs[idx]).collect();
    tensor::normalize_in_place(&mut truncated);
    kept[sample_from_probs(&truncated, state)]
}

/// Categorical sampling via inverse CDF: draw r ~ U(0,1), walk the CDF
/// until cumulative probability exceeds r. Falls back to last token if
/// floating-point rounding causes the walk to not trigger.
fn sample_from_probs(probs: &[f32], state: &mut u64) -> usize {
    let r = next_rand01(state);
    let mut cum = 0.0;
    for (i, p) in probs.iter().enumerate() {
        cum += *p;
        if r <= cum {
            return i;
        }
    }
    probs.len().saturating_sub(1)
}

/// LCG-based PRNG returning a float in [0, 1).
/// Uses the upper 32 bits of the 64-bit state for better quality than lower bits.
fn next_rand01(state: &mut u64) -> f32 {
    *state = state.wrapping_mul(6364136223846793005).wrapping_add(1);
    let hi = (*state >> 32) as u32;
    hi as f32 / 4294967296.0
}

File: julia/generate.jl
# SPDX-License-Identifier: CC-BY-NC-4.0
# Copyright (c) 2025-2026 fumi-engineer

# generate.jl — Text generation with multiple sampling strategies
#
# Implements autoregressive token generation using the MoE Transformer.
# Sampling strategies use Julia's multiple dispatch: pick_token() dispatches
# on the strategy type, so adding a new strategy requires only a new struct
# and a new method — no if/else chains.
#
# PRNG: Uses a simple LCG (Linear Congruential Generator) for deterministic
# reproducibility across languages (same seed -> same sequence in all 4 impls).

abstract type SamplingStrategy end
struct GreedySampling <: SamplingStrategy end
struct TemperatureSampling <: SamplingStrategy
    temperature::Float32
    state::Ref{UInt64}       # mutable PRNG state (LCG)
end
struct TopKSampling <: SamplingStrategy
    k::Int
    temperature::Float32
    state::Ref{UInt64}
end
struct TopPSampling <: SamplingStrategy
    top_p::Float32
    temperature::Float32
    state::Ref{UInt64}
end

# LCG PRNG: state = state * 6364136223846793005 + 1, return upper 32 bits as [0,1)
# Matches the same constants used in the Rust/Go/Python implementations
# for cross-language reproducibility.
function next_rand01(state::Ref{UInt64})
    state[] = state[] * UInt64(6364136223846793005) + UInt64(1)
    Float32(UInt32(state[] >> 32)) / 4294967296f0
end

# Sample from a discrete probability distribution using inverse CDF
function _sample_from_probs(probs::AbstractVector{Float32}, state::Ref{UInt64})
    r = next_rand01(state)
    cum = 0f0
    @inbounds for i in eachindex(probs)
        cum += probs[i]
        r <= cum && return i
    end
    length(probs)
end

# Temperature sampling: scale logits by 1/T, then softmax, then sample
# Temperature=0 falls back to greedy (argmax).
function _sample_from_logits(logits::AbstractVector{Float32}, temperature::Float32, state::Ref{UInt64})
    if temperature <= 0f0
        idx, _ = argmax_f32(logits)
        return idx
    end
    scaled = logits .* (1f0 / temperature)
    softmax_in_place!(scaled)
    _sample_from_probs(scaled, state)
end

# Top-K sampling: keep only the K highest-scoring tokens, mask rest to -inf,
# then apply temperature sampling.
function _sample_topk(logits::AbstractVector{Float32}, k::Int, temperature::Float32, state::Ref{UInt64})
    n = length(logits)
    (k <= 0 || k >= n) && return _sample_from_logits(logits, temperature, state)
    top_idx = partialsortperm(logits, 1:k; rev=true)
    filtered = fill(-Float32(3.4028235e38), n)
    @inbounds for i in top_idx
        filtered[i] = logits[i]
    end
    _sample_from_logits(filtered, temperature, state)
end

# Top-P (nucleus) sampling: keep tokens whose cumulative probability >= top_p,
# then sample from this nucleus set.
function _sample_topp(logits::AbstractVector{Float32}, top_p::Float32, temperature::Float32, state::Ref{UInt64})
    (top_p <= 0f0 || top_p >= 1f0) && return _sample_from_logits(logits, temperature, state)
    if temperature <= 0f0
        idx, _ = argmax_f32(logits)
        return idx
    end
    scaled = logits .* (1f0 / temperature)
    softmax_in_place!(scaled)
    indices = sortperm(scaled; rev=true)
    selected = Int[]
    s = 0f0
    for idx in indices
        push!(selected, idx)
        s += scaled[idx]
        s >= top_p && break
    end
    trunc = Float32[scaled[i] for i in selected]
    normalize_in_place!(trunc)
    chosen = _sample_from_probs(trunc, state)
    selected[chosen]
end

# --- Multiple dispatch: pick_token selects sampling strategy at the type level ---

function pick_token(::GreedySampling, logits::AbstractVector{Float32})
    idx, _ = argmax_f32(logits)
    idx - 1  # 0-based token ID
end

function pick_token(s::TemperatureSampling, logits::AbstractVector{Float32})
    _sample_from_logits(logits, s.temperature, s.state) - 1
end

function pick_token(s::TopKSampling, logits::AbstractVector{Float32})
    _sample_topk(logits, s.k, s.temperature, s.state) - 1
end

function pick_token(s::TopPSampling, logits::AbstractVector{Float32})
    _sample_topp(logits, s.top_p, s.temperature, s.state) - 1
end

# Autoregressive generation loop:
#   1. Forward pass on current token sequence
#   2. Extract logits for the last position
#   3. Sample next token using the chosen strategy
#   4. Append and repeat
# Note: This is a naive implementation that re-computes the full sequence
# each step (no KV cache). Suitable for educational purposes.
function generate(m::MoETransformer, prompt::Vector{Int}, max_len::Int, strategy::SamplingStrategy)
    tokens = copy(prompt)
    isempty(tokens) && push!(tokens, 0)
    while length(tokens) < max_len
        logits = forward_ids(m, tokens, 1, length(tokens))
        last = @view logits.data[1, length(tokens), :]  # logits for last position
        push!(tokens, pick_token(strategy, last))
    end
    tokens
end

# --- Convenience wrappers for common strategies ---

generate_greedy(m::MoETransformer, prompt::Vector{Int}, max_len::Int) =
    generate(m, prompt, max_len, GreedySampling())

generate_sample(m::MoETransformer, prompt::Vector{Int}, max_len::Int,
                temperature::Float32, seed::UInt64) =
    generate(m, prompt, max_len, TemperatureSampling(temperature, Ref(seed)))

generate_topk(m::MoETransformer, prompt::Vector{Int}, max_len::Int,
              k::Int, temperature::Float32, seed::UInt64) =
    generate(m, prompt, max_len, TopKSampling(k, temperature, Ref(seed)))

generate_topp(m::MoETransformer, prompt::Vector{Int}, max_len::Int,
              top_p::Float32, temperature::Float32, seed::UInt64) =
    generate(m, prompt, max_len, TopPSampling(top_p, temperature, Ref(seed)))
